platform/pwl.c

```diff
@@ -39,3 +61,775 @@ pwl_display_destroy (PwlDisplay *self)
+
+    g_assert (wl_data.compositor);
+    g_assert (wl_data.xdg_shell != NULL ||
+              wl_data.shell != NULL ||
+              wl_data.fshell != NULL);
+
+    return TRUE;
+}
+
+void clear_wayland (void)
+{
+    g_source_destroy (wl_data.event_src);
+
+    if (wl_data.xdg_shell != NULL)
+        xdg_wm_base_destroy (wl_data.xdg_shell);
+    if (wl_data.fshell != NULL)
+        zwp_fullscreen_shell_v1_destroy (wl_data.fshell);
+    if (wl_data.shell != NULL)
+        wl_shell_destroy (wl_data.shell);
+
+    if (wl_data.compositor != NULL)
+        wl_compositor_destroy (wl_data.compositor);
+
+    wl_registry_destroy (wl_data.registry);
+    wl_display_flush (wl_data.display);
+    wl_display_disconnect (wl_data.display);
+}
+
+static void
+xdg_surface_on_configure (void *data,
+                          struct xdg_surface *surface,
+                          uint32_t serial)
+{
+    xdg_surface_ack_configure (surface, serial);
+}
+
+static const struct xdg_surface_listener xdg_surface_listener = {
+    .configure = xdg_surface_on_configure
+};
+
+static void
+xdg_toplevel_on_configure (void *data,
+                           struct xdg_toplevel *toplevel,
+                           int32_t width, int32_t height,
+                           struct wl_array *states)
+{
+    configure_surface_geometry (width, height);
+
+    g_debug ("New XDG toplevel configuration: (%" PRIu32 ", %" PRIu32 ")", width, height);
+
+    wl_data.resize_window(data);
+}
+
+static void
+xdg_toplevel_on_close (void *data, struct xdg_toplevel *xdg_toplevel)
+{
+    // TODO g_application_quit (g_application_get_default ());
+}
+
+static const struct xdg_toplevel_listener xdg_toplevel_listener = {
+    .configure = xdg_toplevel_on_configure,
+    .close = xdg_toplevel_on_close,
+};
+
+
+gboolean create_window (void *data, GError **error)
+{
+    g_debug ("Creating Wayland surface...");
+
+    win_data.wl_surface = wl_compositor_create_surface (wl_data.compositor);
+    g_assert (win_data.wl_surface);
+
+#if HAVE_DEVICE_SCALING
+    wl_surface_add_listener (win_data.wl_surface, &win_data.surface_listener, data);
+#endif /* HAVE_DEVICE_SCALING */
+
+    if (wl_data.xdg_shell != NULL) {
+        win_data.xdg_surface =
+            xdg_wm_base_get_xdg_surface (wl_data.xdg_shell,
+                                         win_data.wl_surface);
+        g_assert (win_data.xdg_surface);
+
+        xdg_surface_add_listener (win_data.xdg_surface, &xdg_surface_listener,
+                                  NULL);
+        win_data.xdg_toplevel =
+            xdg_surface_get_toplevel (win_data.xdg_surface);
+        g_assert (win_data.xdg_toplevel);
+
+        xdg_toplevel_add_listener (win_data.xdg_toplevel,
+                                   &xdg_toplevel_listener, data);
+        // TODO: xdg_toplevel_set_title (win_data.xdg_toplevel, COG_DEFAULT_APPNAME);
+
+        const char *app_id = NULL;
+        GApplication *app = g_application_get_default ();
```

Comment:
Setting the application identifier is going to be tricky without breaking the encapsulation. Maybe we'll have to add it as a property of `CogShell`, so then the `CogFdoShell` implementation can call into a `pwl_window_set_application_id()` in turn.

I don't have a very clear idea how to proceed here, so we can leave a `TODO` around for this.